### DISCUSSION TITLE
EG4 Battery Support

### DISCUSSION THREAD

**Discussion by pchiquit** [Louisvdw/dbus-serialbattery#136] [Ideas]
Hi! 
    I've searched through the repo and didn't find any support for this battery "EG4 Lifepower" https://signaturesolar.com/eg4-lifepower4-lithium-battery-24v-200ah/. I'm considering adding the support myself if you are open to a PR with that. I have the battery and would love to be able to get the SoC.
    I'm planning on basing my work on this guy's work: https://github.com/slim-bean/powermon
    Let me know what you think.

**Comment by Louisvdw** [comment]
Hi @pchiquit 
Please do see if you can add it and make a PR. You will see there is a file called battery_template.py for a starting point. Make a copy of it and rename it something like eg4.py and make your commands to read the information. If you can populte the values for the Battery object then the driver will do the rest. You also do not need to have everything polulated.
Have a look at what the other batteries does already. Perhaps one is close to what you need to do so you can just copy and adjust that.

**Comment by pchiquit** [comment]
Hi @Louisvdw . Thanks for the tips. I'm already working on it. One question I have is how to debug this? Do I have to debug on Venus OS or  do you have any framework so I could debug on my personal computer using Visual Studio Code? I was trying to run the script standalone, but it seems to be part of a dbus loop and these python modules seem to be part of Victron. Anyway, I would appreciate any tips you can share on your development workflow.

**Reply by Louisvdw** [reply]
Hi @pchiquit 

I normally start by creating a little test script for the BMS I am working on that I can run on any PC. In the test script I then test all the commands and check the returned values which I then put into the full battery class for that BMS. Attached is an example test script. It use the same serial reader function as the full script, so once all your settings is working you can just move this over. 
[serialtestJBD.zip](https://github.com/Louisvdw/dbus-serialbattery/files/8954985/serialtestJBD.zip)

If you want to test the final changes to dbus-serialbattery you can only run that on a VenusOS device. There is not an easy way to create a virtual enviroment for that. What you can do is to stop the port that your device use from being started by serial-starter and then manually run the driver. 
`/opt/victronenergy/serial-starter/stop-tty.sh ttyUSB0`

And you can start the driver manually then with 
`python /data/etc/dbus-serialbattery/dbus-serialbattery.py /dev/ttyUSB0`
